File: backend/app/schemas/document.py
"""
Document schemas.
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from uuid import UUID
from app.schemas.base import BaseSchema, TimestampMixin, PaginatedResponse


class DocumentLineBase(BaseModel):
    """Base document line fields."""
    line_type: str = Field("product", pattern="^(product|text)$")
    product_id: Optional[UUID] = None
    description: str = Field(..., min_length=1)
    quantity: float = Field(1, gt=0)
    unit_price: float = Field(0, ge=0)
    discount_percent: float = Field(0, ge=0, le=100)


class DocumentLineCreate(DocumentLineBase):
    """Document line creation schema."""
    pass


class DocumentLineResponse(BaseSchema):
    """Document line response."""
    id: UUID
    document_id: UUID
    line_type: str
    product_id: Optional[UUID] = None
    description: str
    quantity: float
    unit_price: float
    discount_percent: float
    subtotal: float
    order_index: int


class DocumentBase(BaseModel):
    """Base document fields."""
    type: str = Field(..., pattern="^(quote|delivery_note|invoice)$")
    client_id: UUID
    issue_date: datetime
    due_date: Optional[datetime] = None
    notes: Optional[str] = None
    internal_notes: Optional[str] = None
    terms: Optional[str] = None


class DocumentCreate(DocumentBase):
    """Document creation with lines."""
    lines: List[DocumentLineCreate] = Field(default_factory=list)


class DocumentUpdate(BaseModel):
    """Document update - limited fields."""
    client_id: Optional[UUID] = None
    due_date: Optional[datetime] = None
    notes: Optional[str] = None
    internal_notes: Optional[str] = None
    terms: Optional[str] = None
    lines: Optional[List[DocumentLineCreate]] = None


class StatusChange(BaseModel):
    """Change document status."""
    new_status: str = Field(
        ...,
        pattern="^(draft|not_sent|sent|accepted|rejected|paid|partially_paid|cancelled)$"
    )


class DocumentResponse(BaseSchema, TimestampMixin):
    """Document response with all details."""
    id: UUID
    company_id: UUID
    code: str
    type: str
    status: str
    issue_date: datetime
    due_date: Optional[datetime] = None
    client_id: UUID
    parent_document_id: Optional[UUID] = None
    subtotal: float
    tax_amount: float
    total: float
    notes: Optional[str] = None
    internal_notes: Optional[str] = None
    terms: Optional[str] = None
    created_by: UUID
    lines: List[DocumentLineResponse] = Field(default_factory=list)


class DocumentSummary(BaseSchema):
    """Document summary for lists."""
    id: UUID
    code: str
    type: str
    status: str
    issue_date: datetime
    due_date: Optional[datetime] = None
    client_id: UUID
    client_name: Optional[str] = None  # Added for dashboard display
    total: float
    created_at: Optional[datetime] = None


class DocumentList(PaginatedResponse[DocumentSummary]):
    """Paginated document list."""
    pass

File: frontend/src/types/document.ts
export interface DocumentLine {
  id: string;
  document_id: string;
  line_type: "product" | "text";
  product_id?: string;
  description: string;
  quantity: number;
  unit_price: number;
  discount_percent: number;
  subtotal: number;
  order_index: number;
  measurements?: string;
}

export interface DocumentLineCreate {
  line_type?: string;
  product_id?: string;
  description: string;
  quantity?: number;
  unit_price?: number;
  discount_percent?: number;
  measurements?: string;
}

export interface Document {
  id: string;
  company_id: string;
  code: string;
  type: string;
  status: string;
  issue_date: string;
  due_date?: string;
  client_id: string;
  parent_document_id?: string;
  subtotal: number;
  tax_amount: number;
  total: number;
  notes?: string;
  internal_notes?: string;
  terms?: string;
  // Business fields
  client_reference?: string;
  payment_method?: string;
  execution_location?: string;
  payment_date?: string;
  created_by: string;
  lines: DocumentLine[];
  created_at: string;
  updated_at: string;
}

export interface DocumentSummary {
  id: string;
  code: string;
  type: string;
  status: string;
  issue_date: string;
  due_date?: string;
  client_id: string;
  client_name?: string;
  total: number;
  created_at?: string;
}

export interface DocumentCreate {
  type: string;
  client_id: string;
  issue_date: string;
  due_date?: string;
  notes?: string;
  internal_notes?: string;
  terms?: string;
  client_reference?: string;
  payment_method?: string;
  execution_location?: string;
  lines: DocumentLineCreate[];
}

export interface DocumentUpdate {
  client_id?: string;
  due_date?: string;
  notes?: string;
  internal_notes?: string;
  terms?: string;
  client_reference?: string;
  payment_method?: string;
  execution_location?: string;
  payment_date?: string;
  lines?: DocumentLineCreate[];
}

export interface DocumentListParams {
  skip?: number;
  limit?: number;
  doc_type?: string;
  doc_status?: string;
  client_id?: string;
  date_from?: string;
  date_to?: string;
  search?: string;
}

export interface StatusChange {
  new_status: string;
}
